File: iOS/Sources/Presentation/ViewModels/AuthViewModel.swift
import Foundation

@MainActor
class AuthViewModel: ObservableObject {
    private let authInteractor: AuthInteractor

    @Published var isLoading = false
    @Published var errorMessage: String?

    var isLoggedIn: Bool {
        authInteractor.isLoggedIn
    }

    var currentUser: User? {
        authInteractor.currentUser
    }

    init(authInteractor: AuthInteractor) {
        self.authInteractor = authInteractor
    }

    func login(email: String, password: String) async {
        isLoading = true
        errorMessage = nil

        do {
            _ = try await authInteractor.login(email: email, password: password)
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    func register(email: String, password: String, name: String) async {
        isLoading = true
        errorMessage = nil

        do {
            _ = try await authInteractor.register(email: email, password: password, name: name)
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    func logout() async {
        await authInteractor.logout()
    }

    func clearError() {
        errorMessage = nil
    }
}

File: android/app/src/main/java/com/budgetmanager/ui/viewmodel/AuthViewModel.kt
package com.budgetmanager.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.budgetmanager.domain.entity.User
import com.budgetmanager.domain.interactor.AuthInteractor
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class AuthUiState(
    val isLoading: Boolean = false,
    val isLoggedIn: Boolean = false,
    val user: User? = null,
    val error: String? = null
)

class AuthViewModel(
    private val authInteractor: AuthInteractor
) : ViewModel() {
    private val _uiState = MutableStateFlow(AuthUiState())
    val uiState: StateFlow<AuthUiState> = _uiState.asStateFlow()

    init {
        checkLoginStatus()
    }

    private fun checkLoginStatus() {
        viewModelScope.launch {
            val isLoggedIn = authInteractor.isLoggedIn()
            val user = if (isLoggedIn) authInteractor.getCurrentUser() else null
            _uiState.value = _uiState.value.copy(
                isLoggedIn = isLoggedIn,
                user = user
            )
        }
    }

    fun login(email: String, password: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)

            authInteractor.login(email, password)
                .onSuccess { response ->
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        isLoggedIn = true,
                        user = response.user,
                        error = null
                    )
                }
                .onFailure { exception ->
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = exception.message ?: "Login failed"
                    )
                }
        }
    }

    fun register(email: String, password: String, name: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)

            authInteractor.register(email, password, name)
                .onSuccess { response ->
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        isLoggedIn = true,
                        user = response.user,
                        error = null
                    )
                }
                .onFailure { exception ->
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = exception.message ?: "Registration failed"
                    )
                }
        }
    }

    fun logout() {
        viewModelScope.launch {
            authInteractor.logout()
            _uiState.value = AuthUiState(isLoggedIn = false)
        }
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }
}
